pkg/operator/bootstrapteardown/waitforceo.go

```diff
@@ -73,3 +84,111 @@ func done(etcd *operatorv1.Etcd) (bool, error) {
+		klog.Errorf("error watching kube-apiservers: %#v", err)
+	}
+	return err
+}
+
+func waitForKASOperator(ctx context.Context, operatorConfig rest.Interface, configMapsGetter v1.ConfigMapsGetter) interface{} {
+	_, err := clientwatch.UntilWithSync(
+		ctx,
+		cache.NewListWatchFromClient(operatorConfig, "kube-apiservers", "", fields.OneTermEqualSelector("metadata.name", "cluster")),
+		&operatorv1.KubeAPIServer{},
+		nil,
+		func(event watch.Event) (bool, error) {
+			switch event.Type {
+			case watch.Added, watch.Modified:
+				apiserver, ok := event.Object.(*operatorv1.KubeAPIServer)
+				if !ok {
+					klog.Warningf("Expected a KubeAPIServer object but got a %q object instead", event.Object.GetObjectKind().GroupVersionKind())
+					return false, nil
+				}
+				return doneApiServer(apiserver, configMapsGetter), nil
+			}
+			klog.Infof("Still waiting for kube-apiserver operator to be healthy...")
+			return false, nil
+		},
+	)
+
+	if err != nil {
+		klog.Errorf("error waiting for kube-apiserver operator: %#v", err)
+		return err
+	}
+
+	klog.Infof("kube-apiserver operator is healthy")
+	return nil
+}
+
+func doneApiServer(kasOperator *operatorv1.KubeAPIServer, configMapsGetter v1.ConfigMapsGetter) bool {
+	revisionMap := map[int32]struct{}{}
+	uniqueRevisions := []int32{}
+
+	for _, nodeStatus := range kasOperator.Status.NodeStatuses {
+		revision := nodeStatus.CurrentRevision
+		if _, ok := revisionMap[revision]; !ok {
+			revisionMap[revision] = struct{}{}
+			uniqueRevisions = append(uniqueRevisions, revision)
+		}
+	}
+
+	// For each revision, check that the configmap for that revision contains the
+	// appropriate storageConfig
+	for _, revision := range uniqueRevisions {
+		configMapNameWithRevision := fmt.Sprintf("%s-%d", configMapName, revision)
+		configMap, err := configMapsGetter.ConfigMaps(operatorclient.TargetNamespace).Get(configMapNameWithRevision, metav1.GetOptions{})
+		if err != nil {
+			klog.Errorf("doneApiServer: error getting configmap: %#v", err)
+			return false
+		}
+		if !configMapHasRequiredValues(configMap) {
+			return false
+		}
+	}
+	return false
+}
+
+type ConfigData struct {
+	StorageConfig struct {
+		Urls []string
+	}
+}
+
+func configMapHasRequiredValues(configMap *corev1.ConfigMap) bool {
+	config, ok := configMap.Data[configMapKey]
+	if !ok {
+		klog.Errorf("configMapHasRequiredValues: config.yaml key missing")
+		return false
+	}
+	var configData ConfigData
+	err := json.Unmarshal([]byte(config), &configData)
+	if err != nil {
+		klog.Errorf("configMapHasRequiredValues: error reading ")
+		return false
+	}
+	if len(configData.StorageConfig.Urls) < 1 {
+		klog.Infof("configMapHasRequiredValues: length of storageUrls %#v less than 1", configData.StorageConfig.Urls)
+		return false
+	}
+	if len(configData.StorageConfig.Urls) == 1 &&
+		!strings.Contains(configData.StorageConfig.Urls[0], "etcd") {
+		klog.Infof("configMapHasRequiredValues: config just has bootstrap IP: %#v", configData.StorageConfig.Urls)
```

Comment:
yes, it would be true, but I am assuming it wouldn't be as disruptive as this. KAS should be pretty stable in that case, it would probably be another iteration on this logic, can add a TODO if it helps...

Reply:
I am saying in restore we could have a` len(configData.StorageConfig.Urls) == 1 ` but not be bootstrap.

Reply:
ohh yes, but the second condition `!strings.Contains(configData.StorageConfig.Urls[0], "etcd")` will not pass in restore so this will not be true?